docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,239 @@
+type Terms struct {
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
+    Collector sdk.AccAddress // address that will collect subscription payments
+}
+```
+
+```go
+// Represents a user's subscription to a service
+type Subscription {
+    Name       string
+    Subscriber sdk.AccAddress
+    Limit      int64
+    LastPaid   time.Time
+}
+```
+
+Message Handling:
+
+Note much of this logic may be done inside a keeper. This is a outline of how msgs will be handled.
+
+```go
+func HandleCreateSubscriptionMsg(ctx sdk.Context, msg CreateSubscriptionMsg) {
+    if SubscriptionExists(msg.Name) {
+        return DuplicateSubscriptionErr
+    }
+    StoreTerms(ctx, msg.Name, msg.Amount, msg.Period, msg.Collector) // Store Terms in store under key "Terms:msg.Name"
+    InitializeDueQueue(ctx, msg.Name) // Initialize empty queue and store under key "DueQueue:Name"
+}
+```
+
+```go
+func HandleSubscribeMsg(ctx sdk.Context, msg SubscribeMsg) {
+    if !SubscriptionExists(msg.Name) {
+        return NoSuchSubscriptionErr
+    }
+
+    subscriptionID := hash(msg.Name|msg.Subscriber)
+    Terms := GetTerms(ctx, msg.Name)
+
+    if msg.Period not in Terms.Periods {
+        return InvalidPeriodErr
+    }
+    
+    // check if user already has subscribe to this service
+    // if so renew the subscription
+    if subscription := GetUserSubscriptions(msg.Subscriber, msg.Name); subscription != nil {
+        if msg.Limit == -1 {
+            subscription.Limit = -1
+        } else {
+            subscription.Limit += msg.Limit
+        }
+        StoreSubscription(subscriptionID, subscription)
+        return nil
+    }
+
+    // new subscribeMsg will pay for the first period
+    err := bank.Send(msg.Subscriber, Terms.Collector, Terms.Amount)
+    if err != nil {
+        return err
+    }
+
+    subscription := Subscription{
+        Name:       Terms.Name,
+        Subscriber: msg.Subscriber,
+        Limit:      msg.Limit,
+        LastPaid:   ctx.BlockTime,
+    }
+    StoreSubscription(subscriptionID, subscription)
+    // adds subscription to list of user subscriptions
+    AppendToUserSubscriptions(subscriptionID)
+    // push to back of DueQueue for this service
+    PushToBackOfDueQueue(msg.Name, subscriptionID)
+}
+```
+
+```go
+// simply deletes mapping subscriptionID => subscription and removes ID from user list
+// will not mutate duequeue
+func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
+    DeleteSubscription(msg.Subscriber, msg.Name)
+}
+```
+
+```go
+func HandleCollectMsg(ctx sdk.Context, msg CollectMsg) {
+    if !SubscriptionExists(msg.Name) {
+        return NoSuchSubscriptionErr
+    }
+    Terms := GetTerms(ctx, msg.Name)
+    if msg.Collector != Terms.Collector {
+        return InvalidCollectorErr
+    }
+    limit := msg.Limit
+    for limit > 0 {
+        for _, period := Terms.Period {
```

Comment:
I'm confused by this logic please add a comment for each loop describing what's happening

Reply:
`CollectMsg` has a `Limit` field to define the maximum number of subscriptions to process in the handling of msg. This is to prevent the issue describe [here](https://github.com/cosmos/cosmos-sdk/issues/4642#issuecomment-513377137)

Each subscription provider maintains a separate `DueQueue` for each period they allow. The reason for this is explained [here](https://github.com/cosmos/cosmos-sdk/pull/4780#discussion_r307527230)

Added comments to loop. Let me know if they aren't sufficient